qutip/core/data/convert.pyx

```diff
@@ -260,61 +262,170 @@ cdef class _to:
+        Python object.
+
+        Parameters
+        ----------
+        creators : iterable of (condition, creator, [priority])
+            An iterable of 2- or 3-tuples describing the new data layer
+            creation functions.
+            Each element can individually be a 2- or 3-tuple; they do not need
+            to be all one or the other.
+
+            Elements
+            ........
+            condition : callable (object) -> bool
+                Function determining if the given object can be converted to a
+                data-layer type using this creator.
+
+            creator function: callable (object, shape) -> Data
+                The creator function. It should take an object and a shape
+                and return a data-layer type instance. The object may be
+                any object for which the condition function returned ``True``
+                when tested.
+
+            priority : real, optional (0)
+                The priority associated with this creator. Higher priority
+                conditions will be tested first and the first valid creator
+                (i.e. for which ``condition(object) == True``) will handle
+                the creation.
+
+        Notes
+        -----
+        Default creators are added with the following priorities:
+
+        * Objects that are instances of data-layer types are
+          converted using ``.copy`` with priority 100.
+        * Objects that have a direct equivalent such as ``numpy.ndarray``
+          or ``scipy.sparse.csr_matrix`` are converted with priority 80.
+        * Objects for which ``scipy.sparse.issparse`` is ``True``
+          are converted using an internal CSR converter with
+          priority 20.
+        * If no condition are meet, ``numpy.array`` is used to try convert
+          the input to an array (priority -inf).
+        """
+        for condition, creator, *priority in creators:
+            if not callable(condition):
+                raise TypeError(str(condition) + " is not a callable")
+            if not callable(create):
+                raise TypeError(str(create) + " is not a callable")
+            priority = float(priority[0] if priority else 0)
+            self._creators.append((condition, creator, priority))
+        self._creators.sort(key=lambda creator: creator[2], reverse=True)
+
+    def __call__(self, arg, shape=None, copy=True):
+        """
+        Build a :class:`qutip.data.Data` object from arg.
+
+        Parameters
+        ----------
+        arg : object
+            Object to be converted to `qutip.data.Data`. Anything that can be
+            converted to a numpy array are valid input.
+
+        shape : tuple, optional
+            The shape of the output as (``rows``, ``columns``).
+        """
+        for condition, create, _ in self._creators:
+            if condition(arg):
+                return create(arg, shape, copy)
```

Comment:
```suggestion
                return create(arg, shape, copy=copy)
```

Since we're relying on the `Dense` and `CSR` constructors, let's pass `copy` as a keyword.  I think `arg` and `shape` are fine to pass positionally, but having `copy`, which is usually an optional argument, be keyword gives user-defined functions a bit more flexibility.